authenticator/server/config/config.go

```diff
@@ -27,194 +125,232 @@ type Config struct {
-	viper.SetConfigType("yaml")
-	viper.AddConfigPath(config.ConfigFilePath)
-
-	// Find and read the config file
-	if err := viper.ReadInConfig(); err != nil {
-		return Config{}, err
+	if err := overrideWithFlagConf(conf); err != nil {
+		return nil, err
 	}
-	if err := viper.Unmarshal(&config); err != nil {
-		return Config{}, err
+	if err := overrideWithEnvConf(conf); err != nil {
+		return nil, err
 	}
-	return config, nil
+	return conf, nil
 }
 
-func verify(config Config) error {
-	if !config.DisableTLS {
-		if config.PathToTLSCert == "" {
-			return errors.New("tls is enabled but no tls cert has been provided")
-		}
-		if config.PathToTLSKey == "" {
-			return errors.New("tls is enabled but no tls key has been provided")
+func defaultConfig() (*Config, error) {
+	conf := &Config{}
+	writableConfFields := reflect.ValueOf(conf).Elem()
+	for _, registeredField := range allRegisteredFields {
+		strDefault := fmt.Sprintf("%v", registeredField.defaultVal)
+		if err := setField(writableConfFields, registeredField, strDefault); err != nil {
+			return nil, err
 		}
 	}
-	return nil
+	return conf, nil
 }
 
-func setConfigEnvVars() error {
-	viper.SetEnvPrefix("approzium")
-	if err := viper.BindEnv("Host", "APPROZIUM_HOST"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("HTTPPort", "APPROZIUM_HTTP_PORT"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("GRPCPort", "APPROZIUM_GRPC_PORT"); err != nil {
-		return err
-	}
+func overrideWithFileConf(conf *Config) error {
+	writableConfFields := reflect.ValueOf(conf).Elem()
 
-	if err := viper.BindEnv("DisableTLS", "APPROZIUM_DISABLE_TLS"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("PathToTLSCert", "APPROZIUM_PATH_TO_TLS_CERT"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("PathToTLSKey", "APPROZIUM_PATH_TO_TLS_KEY"); err != nil {
+	body, err := ioutil.ReadFile(flagConf.ConfigFilePath)
+	if err != nil {
 		return err
 	}
 
-	if err := viper.BindEnv("LogLevel", "APPROZIUM_LOG_LEVEL"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("LogFormat", "APPROZIUM_LOG_FORMAT"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("LogRaw", "APPROZIUM_LOG_RAW"); err != nil {
-		return err
-	}
-	if err := viper.BindEnv("SecretsManager", "APPROZIUM_SECRETS_MANAGER"); err != nil {
+	bodyMap := make(map[section]interface{})
+	if err := yaml.Unmarshal(body, &bodyMap); err != nil {
 		return err
 	}
 
-	if err := viper.BindEnv("ConfigFilePath", "APPROZIUM_CONFIG_FILE_PATH"); err != nil {
-		return err
-	}
+	for section, registeredFields := range fieldRegistry {
 
-	if err := viper.BindEnv("VaultTokenPath", "APPROZIUM_VAULT_TOKEN_PATH"); err != nil {
-		return err
-	}
+		bodyIfc, exists := bodyMap[section]
+		if !exists {
+			continue
+		}
 
-	if err := viper.BindEnv("VaultToken", "VAULT_TOKEN"); err != nil {
-		return err
-	}
+		// Create a map of the fields in the section's body.
+		bodyStr := bodyIfc.(string)
+		bodyStr = strings.ReplaceAll(bodyStr, " = ", "=")
```

Comment:
Why are value assignments in the config file expected to be done via ` = ` instead of the native YAML syntax (`key: val`)?

Reply:
Nice catch! Ugh, I was using a yaml linter that had said that `=` was valid `yaml`, but now I see it should be `:`. Fixing.

Reply:
Fixed in https://github.com/cyralinc/approzium/pull/180/commits/ae2e4a7ce465654c53ac066c09ea0aa65a70ad96